[Sheemo] iP

Sheemo opened the pull request:
# nAkIri 😈
> “Mukachiki! Remembering tasks is so difficult!” – Not a Nakiri Ayame Quote

Greetings Humans!! 'Tis I, [Nakiri Ayame](https://www.youtube.com/channel/UC7fk0CB07ly8oSl0aqKkqFg)! I am a virtual Oni who is more than 1500 years old! You, Humans on the other hand, only live for maybe 100 years. Your life is too short to be remembering tasks and deadlines! Let me help!

Talking to me is easy!

- I understand simple words!
- Use text to chat with me!
- Quick and efficient!

All you need to do is,

1. Download it from [here](https://nus-cs2103-ay2223s2.github.io/website/schedule/week4/project.html).
2. Double-click it.
3. Add your tasks.
4. Let me manage your tasks for you! 😈


Unlike you Humans who care a lot about money, I do all this for ***FREE***! Generous, aren't I? Just be sure to watch my streams!


Features:

- [X] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


For you humans who can code in Java, you can practice Java with me! Here's the `main` method:

```
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Sheemo committed:
Add Greet and echo functionality

Sheemo committed:
Add task adding and listing functionality

Sheemo committed:
Add mark and unmark functionality

Sheemo committed:
Add Deadline, Event and Todo functionality

Sheemo committed:
Add UI testing

Sheemo committed:
Use Exceptions

Sheemo committed:
Add deletion functionality and use enums

Sheemo committed:
Add Save and Load from file functionality

Sheemo committed:
Add time recognition to Deadline and Event

Sheemo committed:
Merge pull request #1 from Sheemo/branch-Level-7

Add Save and Load from file functionality

Sheemo committed:
Merge pull request #2 from Sheemo/branch-Level-8

Add time recognition to Deadline and Event

Sheemo committed:
Split Duke into separate classes

Sheemo committed:
Merge branch 'add-gradle-support' of https://github.com/Sheemo/ip into branch-Level-8

Sheemo committed:
Fix bugs

Sheemo committed:
Merge branch 'master' of https://github.com/Sheemo/ip into branch-Level-8

Sheemo committed:
Add Tests and Create v0.1

Sheemo committed:
Add JavaDocs

Sheemo committed:
Merge pull request #3 from Sheemo/branch-A-JavaDoc

Add JavaDocs

PawanPatil19 commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return Formatted date string according to the above specifications
+     */
+    public String asFileDate() {
```
Perhaps the method name can have a verb?

randallnhr replied:
Agreed, perhaps it could be getFileDate?

PawanPatil19 commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,29 @@
+import duke.TaskList;
+import duke.Ui;
+
+public abstract class Command {
+    protected String cmdWord;
+    private static final Command NONE = new NoCommand();
+
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage);
+
+    public boolean isExit() {
+        return false;
+    }
+
+    private static class NoCommand extends Command {
+        public NoCommand() {
+        }
+
+        @Override
+        public void execute(TaskList tasks, Ui ui, Storage storage) {
+        }
+    }
+
+    public static Command none() {
+        return NONE;
+    }
```
I like the way you have implemented the nested class 👍

PawanPatil19 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,107 @@
+     * @return ArrayList of Task objects that can be passed into a TaskList constructor.
+     * @throws DukeException If the file was corrupted, or could not be read.
+     */
+    public ArrayList<Task> loadTasklistFromFile() throws DukeException {
```
```suggestion
    public ArrayList<Task> loadTaskListFromFile() throws DukeException {
```

PawanPatil19 commented on src/main/java/duke/ErrorEnum.java:
```diff
@@ -0,0 +1,8 @@
+/**
+ * Enum to represent error types. To be deprecated.
+ */
+public enum ErrorEnum {
```
I like the way you have enums for errors 👍

PawanPatil19 reviewed (commented):
LGTM...Just some nits to fix

randallnhr commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Marks a Task as done.
+     */
+    public void markAsDone() {
+        this.isDone = true;
+    }
```
Good job on abtracting the change in isDone boolean to a function, makes the code more readable!

randallnhr commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,66 @@
+ * Driver code for the Duke project.
+ */
+public class Duke {
+    private static final String SAVE_PATH = "./tasklist.txt";
```
Good job in abstracting the file save path! Makes it easier to edit in the future!

dunliang0513 commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return Formatted date string according to the above specifications
+     */
+    public String asFileDate() {
+        return this.isDate()
+                ? this.localDate.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
+                : this.localDateTime.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm"));
+    }
```
Instead of using asFileDate as the method name, I will suggest using method name like formatDate?

dunliang0513 commented on src/main/java/duke/Date.java:
```diff
@@ -0,0 +1,77 @@
+     *
+     * @return Formatted date string according to the above specifications
+     */
+    public String asFileDate() {
```
```suggestion
    public String formatDate() {
```

dunliang0513 commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,22 @@
+package duke;
+
+/**
+ * Parent Exception to all custom exceptions for the Duke project.
+ */
+public class DukeException extends Exception {
+    /**
+     * Constructor for a DukeException.
+     */
+    public DukeException() {
+        super();
+    }
+
+    /**
+     * Constructor for a DukeException with an optional message.
+     *
+     * @param message Exception message.
+     */
+    public DukeException(String message) {
+        super(message);
+    }
+}
```
I feel like you can put all the exceptions(DukeExeption, EmptyBodyException, InvalidIndexException) under a package called DukeMainExceptions.

dunliang0513 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,39 @@
+package duke;
+
+/**
+ * Class that can store a Deadline task, which has an additional "by" field.
+ */
+public class Deadline extends Task {
+    protected Date by;
+
+    /**
+     * Constructor for a Deadline object.
+     *
+     * @param description Task description
+     * @param by Task deadline
+     */
+    public Deadline(String description, Date by) {
+        super(description);
+        this.by = by;
+    }
+
+    /**
+     * Returns a String representation of the task that is parsable by the Duke Storage parser.
+     *
+     * @return A String object that is parsable by the Duke Storage parser.
+     */
+    @Override
+    public String getFileRepresentation() {
+        return "deadline|" + super.getFileRepresentation() + "|" + this.by.asFileDate();
+    }
+
+    /**
+     * Returns a String representation of the task to be printed on the screen.
+     *
+     * @return String representation for a Deadline object.
+     */
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + this.by + ")";
+    }
+}
```
I think can create another package call Tasks which contains Dealine.java, Event.java, Task.java and Todo.java

dunliang0513 reviewed (commented):
Overall, the way you handle the logic and flow is pretty good.
I have provided a few personal suggestions which I think will be helpful for you.
The duke chatbot looks like having a very interesting personality. Well done!

Sheemo committed:
Add Find functionality

Sheemo committed:
Merge pull request #4 from Sheemo/branch-Level-9

Add Find functionality

Sheemo committed:
Add GUI

Sheemo committed:
Merge pull request #6 from Sheemo/branch-Level-10

Add GUI

Sheemo committed:
Add more Exceptions

There was only one exception, which was unclear which error is being handled.

More exceptions were added to have more flexibility in dealing with different types of errors.

Separating exceptions in this way also enables standardization of error messages.

Sheemo committed:
Remove Ui.java

The code became redundant with the addition of the GUI

We can safely delete Ui.java and refactor all relevant code to not use it.

Sheemo committed:
Refactor code

Code was refactored to follow more closely to code quality guidelines.

Code now adheres more to SLAP principle and more abstractions were used. This makes the code more readable.

Sheemo committed:
Improve UI

Sheemo committed:
Merge pull request #7 from Sheemo/branch-A-CodeQuality

Improve UI

Sheemo committed:
Add Priorities to Tasks

Sheemo committed:
Merge pull request #8 from Sheemo/branch-A-Assertions

Add Priorities to Tasks
